db/dbformat.h

```diff
@@ -77,6 +77,30 @@ enum ValueType : unsigned char {
 extern const ValueType kValueTypeForSeek;
 extern const ValueType kValueTypeForSeekForPrev;
 
+// A range of user keys used internally by RocksDB. Also see `Range` used by
+// public APIs.
+struct UserKeyRange {
+  // In case of user_defined timestamp, if enabled, `start` and `limit` should
+  // include user_defined timestamps.
+  Slice start;
+  Slice limit;
+
+  UserKeyRange() {}
+  UserKeyRange(const Slice& s, const Slice& l) : start(s), limit(l) {}
+};
+
+// A range of user keys used internally by RocksDB. Also see `RangePtr` used by
+// public APIs.
+struct UserKeyRangePtr {
+  // In case of user_defined timestamp, if enabled, `start` and `limit` should
+  // point to key with timestamp part.
+  const Slice* start;
+  const Slice* limit;
```

Comment:
The idea behind `UserKeyRangePtr` is that each boundary can be optional, right? If so, we could use `std::optional<Slice>` instead of `const Slice*`

Reply:
That's a good point. Using a pointer type here is really an implicit way of expressing the optional nature of the bound. I have updated their types.